src/Nix/Builtins.hs

```diff
@@ -763,10 +784,15 @@ functionArgs fun = fun >>= \case
 
 toFile :: MonadNix e m => m (NValue m) -> m (NValue m) -> m (NValue m)
 toFile name s = do
-    name' <- fromValue name
+    name' <- fromStringNoContext =<< fromValue name
     s' <- fromValue s
-    mres <- toFile_ (Text.unpack name') (Text.unpack s')
-    toNix $ Text.pack $ unStorePath mres
+    -- TODO Using hacky here because we still need to turn the context into
+    -- runtime references of the resulting file.
+    -- See prim_toFile in nix/src/libexpr/primops.cc
+    mres <- toFile_ (Text.unpack name') (Text.unpack $ hackyStringIgnoreContext s')
+    let t = Text.pack $ unStorePath mres
+        sc = StringContext t DirectPath
+    toNix $ principledMakeNixStringWithSingletonContext t sc
```

Comment:
Might be nice to have a function that takes a path and returns the string with singleton context corresponding to that path... later :smile: